enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,774 @@
+and attempts will be made to migrate working data to newer version before starting MicroShift.
+If new commit is unhealthy, red script will persist "restore" action, thus on next boot to new commit,
+MicroShift's data compatible with old commit will be restored resulting in new attempt to
+migrate the data before starting MicroShift.
+Such approach will allow to attempt data migration procedure multiple times and protect MicroShift
+from being unhealthy due to failed migration.
+It is at a cost of potentially loosing some data that might be produced between MicroShift start and
+host deemed unhealthy by greenboot.
+
+Functionality will be implemented by placing in `/etc/greenboot/green.d` and `/etc/greenboot/red.d`
+bash scripts containing with simple logic or, if needed,
+executing commands `microshift greenboot green` and `microshift greenboot red` in case of needing to put more information into file with action for next boot.
+Alternatively, a symlinks to `microshift` binary can be made and MicroShift modified to run specific command depending on content of argv[0] (just like BusyBox).
+
+File containing said information should not be part of MicroShift data directory
+as we don't to it be a part of backup.
+
+### Backup and restore of MicroShift data
+
+_TODO: 1 backup per ostree deployment_
+
+Decision whether to backup or restore is based on file persisted during previous boot (see "Integration with greenboot").
+When file is read to make a decision (but not during dry run), it shall be removed.
+
+In case of unhealthy first ostree commit there might occur situation that MicroShift upon start is instructed
+to restore the backup which doesn't exist. In such case MicroShift will delete its data directory to
+perform a clean start.
+
+As a result of investigation and aiming for simplicity for initial implementation,
+it was decided that backing up MicroShift's data will be done by leveraging using copy-on-write (CoW) functionality.
+
+CoW is a feature of filesystem (supported by XFS and Btrfs) and it can be used by providing a `--reflink=` param to `cp` option.
+`--reflink=auto` will be used over `--reflink=always` to gracefully fall back to regular copying on filesystems
+not supporting CoW (ext4, ZFS).
+Backup will be done in `/var/lib/microshift.bak` or similar - it needs to be within the same filesystem/mount.
+Only one backup will be kept. Consequent boots will overwrite the backup dir (only if the action is "backup").
+
+Restore is analogous operation, just in different direction - copying contents of `/var/lib/microshift.bak` to `/var/lib/microshift`.
+
+End user documentation needs to include guidance on setting up filesystem to fullfil requirements for using copy-on-write (e.g. making sure some filesystem options are not disabled).
+
+### Contents of MicroShift data backup
+
+- Etcd database shall be backed up fully, rather just creating a snapshot to preserve
+  history and other data that could be lost.
```

Comment:
And also so we don't have to worry about whether the "new" tool is compatible with the existing file format when taking the backup.

Reply:
True, but it needs to be able to migrate the data, so actually it must be compatible